repo2docker/buildpacks/julia/julia_require.py

```diff
@@ -0,0 +1,170 @@
+            ('JUPYTER', '${NB_PYTHON_PREFIX}/bin/jupyter')
+        ]
+
+    def get_path(self):
+        """Adds path to Julia binaries to user's PATH.
+
+         Returns:
+             an ordered list of path strings. The path to the Julia
+             executable is added to the list.
+
+        """
+        return super().get_path() + ['${JULIA_HOME}']
+
+    def get_build_scripts(self):
+        """
+        Return series of build-steps common to "ALL" Julia repositories
+
+        All scripts found here should be independent of contents of a
+        particular repository.
+
+        This creates a directory with permissions for installing julia packages
+        (from get_assemble_scripts).
+
+        """
+        return super().get_build_scripts() + [
+            (
+                "root",
+                r"""
+                mkdir -p ${JULIA_PATH} && \
+                curl -sSL "https://julialang-s3.julialang.org/bin/linux/x64/${JULIA_VERSION%[.-]*}/julia-${JULIA_VERSION}-linux-x86_64.tar.gz" | tar -xz -C ${JULIA_PATH} --strip-components 1
+                """
+            ),
+            (
+                "root",
+                r"""
+                mkdir -p ${JULIA_PKGDIR} && \
+                chown ${NB_USER}:${NB_USER} ${JULIA_PKGDIR}
+                """
+            ),
+            (
+                "${NB_USER}",
+                # HACK: Can't seem to tell IJulia to install in sys-prefix
+                # FIXME: Find way to get it to install under /srv and not $HOME?
+                r"""
+                julia -e 'if (VERSION > v"0.7-") using Pkg; else Pkg.init(); end; Pkg.add("IJulia"); using IJulia;' && \
+                mv ${HOME}/.local/share/jupyter/kernels/julia-${JULIA_VERSION%[.-]*}  ${NB_PYTHON_PREFIX}/share/jupyter/kernels/julia-${JULIA_VERSION%[.-]*}
+                """
+            )
+        ]
+
+    def get_assemble_scripts(self):
+        """
+        Return series of build-steps specific to "this" Julia repository
+
+        Precompile all Julia libraries found in the repository's REQUIRE
+        file. The parent, CondaBuildPack, will add the build steps for
+        any needed Python packages found in environment.yml.
+
+        """
+        require = self.binder_path('REQUIRE')
+        return super().get_assemble_scripts() + [(
+            "${NB_USER}",
+            # Install and pre-compile all libraries if they've opted into it.
+            # In v0.6, Pkg.resolve() installs all the packages, but in v0.7+, we
+            # have to manually Pkg.add() each of them (since the REQUIRES file
+            # format is deprecated).
+            # The precompliation is done via `using {libraryname}`.
+            r"""
+            julia /tmp/install-repo-dependencies.jl "%(require)s"
+            """ % {"require": require}
+            # TODO: For some reason, `rm`ing the file fails with permission denied.
+            # && rm /tmp/install-repo-dependencies.jl
```

Comment:
Sorry I haven't been following all the discussion on this PR. Was anyone ever able to resolve this? There was some weird permissions thing happening where I was getting a permissions error when i tried to clean up this file.

Now that I look closely, I'm guessing it's because this command was running as `"${NB_USER}",` up above on line 138? 

Maybe if you could make this a separate step it would work? Something like:
```suggestion
           ,
           "root",  # OR whatever this is supposed to be?
           "rm /tmp/install-repo-dependencies.jl"
```
? But I actually don't know what the right user would be. Whatever it is in the `get_build_script_files` below?

Reply:
This is unchanged from the version on `master`. I would suggest that if someone comes up with a better solution, it can come in a separate PR.

Reply:
Agree on punting.

---

How does `/tmp/install-repo-dependencies.jl` come to exist? My guess would be that it gets copied here by `root` and that is why NB_USER can't delete it. So I'd suggest we change the UID of the user who creates this file.